### DISCUSSION TITLE
KeyError: '\n' when running on Windows

### DISCUSSION THREAD

**Discussion by ghost** [dlebech/lyrics-generator#35] [Q&A] [answered]
Traceback (most recent call last):
  File "C:\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\Ryan Schatz\Desktop\homework\ai git\lyrics-generator-master\lyrics\train.py", line 483, in <module>
    train(
  File "C:\Users\Ryan Schatz\Desktop\homework\ai git\lyrics-generator-master\lyrics\train.py", line 269, in train
    X, y, seq_length, num_words, tokenizer = prepare_data(
  File "C:\Users\Ryan Schatz\Desktop\homework\ai git\lyrics-generator-master\lyrics\train.py", line 77, in prepare_data
    newline_int = tokenizer.word_index["\n"]
KeyError: '\n'

I keep getting this error no matter what dataset I run. I even tried the sample command and it returned this error.

**Comment by dlebech** [comment, marked as answer]
Hi @mineryan1, This might be a Windows issue. Some background below.

The error you get is because the code expects your input texts to have the newline character `\n` in it to delimit lines. So if your dataset consists of only texts without newlines, then that would explain the error.

When you say you ran the "sample command", do you mean the command:
```shell
python -m lyrics.train --songdata-file sample.csv --early-stopping-patience 50 --artists '*'
```
If you ran that and got the above error, I assume this is a problem with newline endings on Windows (based on your stacktrace, it looks like you're on Windows). I'm on Linux myself, and haven't tested that the code works in Windows.

When viewing the [`sample.csv` file](https://raw.githubusercontent.com/dlebech/lyrics-generator/master/sample.csv), it should look something like this:

![image](https://user-images.githubusercontent.com/292796/119358356-d17ba100-bca8-11eb-9b03-ee3764c9036e.png)

If it doesn't look like that in your text editor, I suspect that's the problem.

You could try to replace `tokenizer.word_index["\n"]` with `tokenizer.word_index[os.linesep]` and see if that works for your own dataset. It will not work for the sample csv file I guess, unless you change the line separator yourself.

Hope this helps :slightly_smiling_face:
